### PR TITLE
Added pre-commit hook

### PR DESCRIPTION
**Pull request by bugnano**
This PR adds a git pre-commit hook that automatically formats all dart files

### CODE REVIEW

**Review by vickz84259 — CHANGES_REQUESTED** [review]
I'd prefer we not have node as a dev dependency on the flutter project. Have you had a look at this: https://github.com/hyiso/lint_staged as an alternative?

**Comment by bugnano** [comment]
Looks like using https://github.com/hyiso/lint_staged is working (almost).

I say almost because I don't fully understand how it works with partially committed files and/or files outside the main `lib/` directory (that is, the files in the `example/push_notifications` directory).

Let me explain:
For local development, I'm changing the files `chatbox.dart`, `conversationlist.dart`, and in the example `main.dart`, to point to the local server and use the correct app id.

When I make some changes to those files, I only partially stash those files, leaving out the parts for local development.

So, I tried to make some random changes to those files, and the ones in the `lib/` directory looked like they worked correctly, but if I tried to stage the main.dart file in the example, lint_staged errored out with the error:

```
✔ Prepared lint_staged                    53ms
✔ Hide unstaged changes                   4ms
✔ Run tasks for staged files              9832ms
✔ Apply modifications                     7ms
✔ Restore unstaged changes                11ms
✔ Revert because of errors                73ms
✔ Cleanup temporary files                 21ms
✗ lint_staged failed due to a git error.
  Any lost modifications can be restored from a git stash:
    > git stash list
    stash@{0}: automatic lint_staged backup
    > git stash apply --index stash@{0}
```

Despite what the output says, there's nothing in the stash, and I cannot find a way to make this more verbose, for understanding what went wrong.

**Side note:**

Now you see in the `pubspec.yaml` file the following:

```yaml
lint_staged:
  'lib/**.dart': dart format --fix && dart fix --apply
```

This is because formatting the files in the `lib` directory seems to work.

My tests, however were made with the following part in `pubspec.yaml`:

```yaml
lint_staged:
  '**/*.dart': dart format --fix && dart fix --apply
```

I also tried:

```yaml
lint_staged:
  'lib/**.dart': dart format --fix && dart fix --apply
  'example/push_notifications/lib/**.dart': dart format --fix && dart fix --apply
```

with the same results.

**Comment by bugnano** [comment]
OK, I managed to get verbose output from `lint_staged`:

```
⠋ Restore unstaged changes...
lit_staged:git - git apply -v --whitespace=nowarn --recount --unidiff-zero .git/lint_staged_unstaged.patch
Checking patch example/push_notifications/lib/main.dart...
error: while searching for:
    final session = Session(appId: 'YOUR_APP_ID', enablePushNotifications: true);

error: patch failed: example/push_notifications/lib/main.dart:37
error: example/push_notifications/lib/main.dart: patch does not apply
lint_staged:workflow - Error while restoring changes:
lint_staged:workflow - Retrying with 3-way merge
lit_staged:git - git apply -v --whitespace=nowarn --recount --unidiff-zero --3way .git/lint_staged_unstaged.patch
Checking patch example/push_notifications/lib/main.dart...
error: while searching for:
    final session = Session(appId: 'YOUR_APP_ID', enablePushNotifications: true);

error: patch failed: example/push_notifications/lib/main.dart:37
Falling back to three-way merge...
Applied patch to 'example/push_notifications/lib/main.dart' with conflicts.
Applied patch example/push_notifications/lib/main.dart cleanly.
U example/push_notifications/lib/main.dart
lint_staged:workflow - Error while restoring unstaged changes using 3-way merge:
lint_staged:workflow - ProcessException: git apply -v --whitespace=nowarn --recount --unidiff-zero --3way .git/lint_staged_unstaged.patch
Checking patch example/push_notifications/lib/main.dart...
error: while searching for:
    final session = Session(appId: 'YOUR_APP_ID', enablePushNotifications: true);

error: patch failed: example/push_notifications/lib/main.dart:37
Falling back to three-way merge...
Applied patch to 'example/push_notifications/lib/main.dart' with conflicts.
Applied patch example/push_notifications/lib/main.dart cleanly.
U example/push_notifications/lib/main.dart
  Command: git apply -v --whitespace=nowarn --recount --unidiff-zero --3way .git/lint_staged_unstaged.patch
lint_staged:workflow - #0      Git.run (package:lint_staged/src/git.dart:55:7)
<asynchronous suspension>
#1      Workflow.resotreUnstagedChanges (package:lint_staged/src/workflow.dart:232:9)
<asynchronous suspension>
#2      runAll (package:lint_staged/src/run.dart:122:5)
<asynchronous suspension>
#3      lintStaged (package:lint_staged/lint_staged.dart:26:17)
<asynchronous suspension>
#4      main (file:///home/fri/.pub-cache/hosted/pub.dev/lint_staged-0.4.0/bin/lint_staged.dart:25:18)
<asynchronous suspension>
```

The error has to do with the fact that the file has been only partially staged, and the line

```dart
    final session = Session(appId: 'YOUR_APP_ID', enablePushNotifications: true);
```

does not exist in my working copy.

FWIW I originally chose the Nodejs version of `lint_staged` because it explicitly supports partially staged files.

**Comment by bugnano** [comment]
Update to this whole `lint_staged` saga:

I opened a bug in the `lint_staged` repo, to see if they can fix it ( https://github.com/hyiso/lint_staged/issues/8 ).

I tried the same test with the JS version of `lint_staged`, and it fails in the same way, so it's safe to assume that what I'm doing is unsupported, and the current implementation of `lint_staged` is good enough for us.

@vickz84259 @moubaraks If that's OK for you, I propose to merge this PR.

File: pubspec.yaml [hunk 1]
```diff
@@ -28,6 +28,11 @@ dependencies:
 dev_dependencies:
   flutter_test:
     sdk: flutter
+  husky: ^0.1.6
+  lint_staged: ^0.4.0
+
+lint_staged:
+  '**/*.dart': dart format --fix && dart fix --apply
```
Comment (vickz84259) [inline pubspec.yaml:35]:
```suggestion
  '**/*.dart': dart format && dart fix --apply
```
Looking at the options available to `dart format`, --fix is not listed.

**Review by vickz84259 — CHANGES_REQUESTED** [review]
One minor fix. Everything else looks good.

File: pubspec.yaml (hunk 1 above)
Comment (bugnano) [reply pubspec.yaml]: Done👍

**Pull request merged**